common/lib/dependabot/ecosystem.rb

```diff
@@ -0,0 +1,246 @@
+        @supported_versions = T.let(supported_versions, T::Array[Dependabot::Version])
+      end
+
+      # The name of the package manager or language (e.g., "bundler", "ruby").
+      # @example
+      #   name #=> "bundler"
+      sig { returns(String) }
+      attr_reader :name
+
+      # The current version of the package manager or language.
+      # @example
+      #   version #=> Dependabot::Version.new("2.1.4")
+      sig { returns(Dependabot::Version) }
+      attr_reader :version
+
+      # The raw current version of the package manager or language as a string.
+      # @example
+      #   raw_version #=> "2.1.4"
+      sig { returns(String) }
+      attr_reader :raw_version
+
+      # The version requirements (optional).
+      # @example
+      #   requirement #=> Requirement.new(">= 2.1")
+      sig { returns(T.nilable(Requirement)) }
+      attr_reader :requirement
+
+      # Returns an array of deprecated versions of the package manager.
+      # @example
+      #  deprecated_versions #=> [Version.new("1")]
+      sig { returns(T::Array[Dependabot::Version]) }
+      attr_reader :deprecated_versions
+
+      # Returns an array of supported versions of the package manager.
+      sig { returns(T::Array[Dependabot::Version]) }
+      attr_reader :supported_versions
+
+      # Checks if the current version is deprecated.
+      # Returns true if the version is in the deprecated_versions array; false otherwise.
+      # @example
+      #   deprecated? #=> true
+      sig { returns(T::Boolean) }
+      def deprecated?
+        return false if unsupported?
+
+        deprecated_versions.include?(version)
+      end
+
+      # Checks if the current version is unsupported.
+      # @example
+      #   unsupported? #=> false
+      sig { returns(T::Boolean) }
+      def unsupported?
+        return false if supported_versions.empty?
+
+        # Check if the version is not supported
+        supported_versions.all? { |supported| supported > version }
+      end
+
+      # Raises an error if the current package manager or language version is unsupported.
+      # If the version is unsupported, it raises a ToolVersionNotSupported error.
+      sig { void }
+      def raise_if_unsupported!
+        return unless unsupported?
+
+        # Example: v2.*, v3.*
+        supported_versions_message = supported_versions.map { |v| "v#{v}.*" }.join(", ")
+
+        raise ToolVersionNotSupported.new(
+          name,
+          version.to_s,
+          supported_versions_message
+        )
+      end
+
+      # Indicates if the package manager supports later versions beyond those listed in supported_versions.
+      # By default, returns false if not overridden in the subclass.
+      # @example
+      #   support_later_versions? #=> true
+      sig { returns(T::Boolean) }
+      def support_later_versions?
+        false
+      end
+    end
+
+    class Requirement
```

Comment:
Can we reuse the existing Dependabot::Requirement?

Also this seems like just a container for member variables but has no functionality. I would prefer if the `initializer` just took the constraint as a string, and then had a `min_version` and `max_version`, etc.

Reply:
Actually the purpose of this PR was converting the structure without making functional changes. For parsing requirement constraints there will be another PR I will consider how to do it since there will be differences between ecoystem parsings. So I am going to remove the `Requirement` since currently it is a null object.